op-node/rollup/driver/sequencer.go

```diff
@@ -12,55 +12,293 @@ import (
+}
+
 // Sequencer implements the sequencing interface of the driver: it starts and completes block building jobs.
 type Sequencer struct {
 	log    log.Logger
 	config *rollup.Config
+	mode   SequencerMode
 
 	engine derive.ResettableEngineControl
 
 	attrBuilder      derive.AttributesBuilder
 	l1OriginSelector L1OriginSelectorIface
+	espresso         EspressoIface
 
 	metrics SequencerMetrics
 
 	// timeNow enables sequencer testing to mock the time
 	timeNow func() time.Time
 
 	nextAction time.Time
+
+	// The current Espresso block we are building, if applicable.
+	espressoBatch *InProgressBatch
 }
 
-func NewSequencer(log log.Logger, cfg *rollup.Config, engine derive.ResettableEngineControl, attributesBuilder derive.AttributesBuilder, l1OriginSelector L1OriginSelectorIface, metrics SequencerMetrics) *Sequencer {
+func NewSequencer(log log.Logger, cfg *rollup.Config, engine derive.ResettableEngineControl, attributesBuilder derive.AttributesBuilder, l1OriginSelector L1OriginSelectorIface, espresso EspressoIface, metrics SequencerMetrics) *Sequencer {
 	return &Sequencer{
 		log:              log,
 		config:           cfg,
+		mode:             Unknown,
 		engine:           engine,
 		timeNow:          time.Now,
 		attrBuilder:      attributesBuilder,
 		l1OriginSelector: l1OriginSelector,
+		espresso:         espresso,
 		metrics:          metrics,
+		espressoBatch:    nil,
 	}
 }
 
-// StartBuildingBlock initiates a block building job on top of the given L2 head, safe and finalized blocks, and using the provided l1Origin.
-func (d *Sequencer) StartBuildingBlock(ctx context.Context) error {
+// startBuildingEspressoBatch initiates an Espresso block building job on top of the given L2 head,
+// safe and finalized blocks. After this function succeeds, `d.espressoBatch` is guaranteed to be
+// non-nil.
+func (d *Sequencer) startBuildingEspressoBatch(ctx context.Context, l2Head eth.L2BlockRef) error {
+	windowStart := l2Head.Time + d.config.BlockTime
+	windowEnd := windowStart + d.config.BlockTime
+
+	// Fetch the available HotShot blocks from this sequencing window. The first block in the window
+	// tells us what L1 origin we're going to be building for.
+	blocks, fromBlock, err := d.espresso.FetchHeadersForWindow(ctx, windowStart, windowEnd)
+	if err != nil {
+		return err
+	}
+	if len(blocks) < 2 {
+		// The first block should be one before the start of the window, as proof that we are
+		// including all the blocks in the window. In order to create the payload attributes, we
+		// need a second block which falls in the window to tell us the L1 origin. If we don't have
+		// that, return a temporary error so we try again shortly.
+		return derive.NewTemporaryError(fmt.Errorf("not enough blocks available to determine L1 origin of next L2 batch"))
+	}
+
+	batch := &InProgressBatch{
+		onto:        l2Head,
+		windowStart: windowStart,
+		windowEnd:   windowEnd,
+	}
+
+	if fromBlock == 0 && blocks[0].Timestamp >= windowStart {
```

Comment:
If the espresso client isn't a trusted entity, I see why these checks are necessary actually

Reply:
It's actually the other way around: if `blocks[0].Timestamp > windowStart`, `fromBlock` must be 0, otherwise we could be missing a block. But if `fromBlock` is 0, it is still possible that `blocks[0]` is before the window and `blocks[1]` is in the window. But in any case, the only time this exception is allowed is when both of these are true so it doesn't hurt to double check